File: src/client/java/com/boclips/users/client/model/contract/Contract.java
package com.boclips.users.client.model.contract;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import lombok.Data;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = SelectedCollectionsContract.class, name = "SelectedCollections"),
        @JsonSubTypes.Type(value = SelectedVideosContract.class, name = "SelectedVideos")
})
@Data
public abstract class Contract {
    protected String name;
}

File: src/main/kotlin/com/boclips/users/domain/model/contract/Contract.kt
package com.boclips.users.domain.model.contract

sealed class Contract {
    data class SelectedCollections(
        override val id: ContractId,
        override val name: String,
        val collectionIds: List<CollectionId>
    ) : Contract()

    data class SelectedVideos(
        override val id: ContractId,
        override val name: String,
        val videoIds: List<VideoId>
    ) : Contract()

    abstract val id: ContractId
    abstract val name: String
}